docs/decisions/0007-lti-1p3-decoupling.rst

```diff
@@ -0,0 +1,115 @@
+get the course key from the location field
+<https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/plugin/views.py#L181>`_, etc. This prevents LTI
+1.3 launches from running outside of the platform.
+
+Open edX would like to support the following LTI use cases:
+
+* content launches in an XBlock context
+* content launches from multiple XBlocks, reusing LTIConfigurations between them (i.e. course context)
+* content launches outside an XBlock context in the platform (e.g. LTI course tabs, LTI 1.3 embedded discussions)
+* content launches from an independently deployable application (i.e. IDA or Django context)
+
+https://github.com/openedx/xblock-lti-consumer/pull/254 decoupled the LTI 1.3 launch handler endpoint
+``launch_gate_endpoint``, the LTI 1.3 keyset endpoint ``public_keyset_endpoint``, and the LTI 1.3 access token endpoint
+``access_token_endpoint`` from the XBlock by moving these endpoints into the Django plugin. This was the first step in
+decoupling the XBlock from the LTI 1.3 launch. This decision describes how to fully decouple the LTI 1.3 launch from the
+XBlock.
+
+Decision
+********
+
+* We will remove references to the ``location`` field from all views that implement the generic LTI 1.3 launch. This
+  includes the ``launch_gate_endpoint``, ``public_keyset_endpoint``, and ``access_token_endpoint`` views.
+* The responsibility of determining the launch information included in an LTI launch (e.g. roles and context claims)
+  will be that of the context in which the LTI 1.3 launch occurs (e.g. XBlock, IDA).
+* The library will expose a function that will return the URL that initiates the LTI 1.3 launch. The function will have
+  a parameter ``claims``, which will be an object that represents the claims that should be included in the LTI 1.3
+  launch.
+* Each context will supply the launch information to the LTI 1.3 launch by including it as an argument to the
+  aforementioned function. Each context will call this function to fetch the URL to trigger an LTI 1.3 launch.
+* In order for this launch information to be available throughout the LTI 1.3 launch, the library will include the
+  launch information provided by the context in the third-party initiated login request through the ``lti_message_hint``
+  query or form parameter.
+* The third-party initiated login request ``login_hint`` query or form parameter will be used to include the user
+  identifier. The ``login_hint`` will not be used to identify the ``LtiConfiguration`` instance. Information necessary
+  to identify the ``LtiConfiguration`` instance will be included in the ``lti_message_hint`` query or form parameter
+  instead.
+* We will not describe how to decouple LTI Advantage Services in this document. This may be done at a later time.
+
+Consequences
+************
+
+* LTI 1.3 launches will be decoupled from the XBlock and from the edX platform. This will allow Open edX to better
```

Comment:
I haven't been able to find a definitive answer to this question, but I'm concerned about the URL length for GET requests if we stuff all this information into the query parameter. We may need to enforce that contexts send POSTs to the LTI 1.3 launch initiation URL instead of GETs, which is a consequence. [Here](https://stackoverflow.com/a/417184) is some advice on the topic.

Reply:
this makes sense, although I worry that we will run into an external tool which demands GET.

Reply:
@MichaelRoytman We can't choose what method tools will use to perform the LTI launch, so we can go with the worst-case scenario here and model our parameters to fit that size.

Instead of using a JWT we can try using something smaller with small parameter names?

Reply:
I hadn't considered how large the parameter count could theoretically grow here. However, most fields could be looked back up again based on the configuration so any 'launch fields' (such as the attempt id for proctoring) should be pretty limited right?

If we're still uncomfortable with how this would scale I might suggest an alternative. We persist a server side 'launch' object and instead of encoding it in the ``lti_message_hint`` and we simply send along a lookup key instead. This would be pretty straightforward to do with a simple cache. It isn't too different from the suggested approach so we could always see if we run into trouble and pivot later.

Reply:
@Zacharis278 :+1: for storing a server-side object with all the launch parameters and just passing a token/identifier through the launch message.

Reply:
What are you referring to as the "launch model"? Do you mean the set of launch data that is defined in the context and passed by the context to the library's API methods?

What is the potential schema of an `LtiPlacement`? I understand it would store a `resource_link_id`, right? I suppose I'm trying to understand the difference between what is in `launch_data` and what would be in `LtiPlacement`.

Just taking a cursory look through the various claims, I could see `resource_link_id` and `launch_presentation` as placement specific values. What about `context`? We could share `LtiConfigurations` across `contexts` (e.g. courses), so this could be considered a placement-specific value. So, where would we draw the line between what's in `launch_data` and what's in an `LtiPlacement`?

I'm not as familiar with deep linking yet. What deep linking configuration could go in `LtiPlacement`?

Reply:
>Just taking a cursory look through the various claims, I could see resource_link_id and launch_presentation as placement specific values. What about context? We could share LtiConfigurations across contexts (e.g. courses), so this could be considered a placement-specific value. So, where would we draw the line between what's in launch_data and what's in an LtiPlacement?

I think that context should come from the placement, no? Each placement would represent a different XBlock being used somewhere in the platform and should yield different contexts.

>So, where would we draw the line between what's in launch_data and what's in an LtiPlacement?

Yeah, this is still a bit confusing - I haven't thought through this all year, just wanted to expose one of the options I originally came up with for consideration.

>I'm not as familiar with deep linking yet. What deep linking configuration could go in LtiPlacement?

None, but each `LtiDlContentItem` would have a Foreign Key to a placement instead of to the `LtiConfig`. Currently there's nothing differentiating between LTI DL items on each placement, meaning they won't work properly if we try to re-use configuration (e.g. with the current setup, any LTI DL configuration that is done for a block will be replicated to any other block using the same configuration, because `LtiDlContentItem` is tied to `LtiConfig`). 

Let me know if that is not clear - I can elaborate.